runtime/src/bank.rs

```diff
@@ -808,6 +829,9 @@ pub struct Bank {
     epoch_reward_status: EpochRewardStatus,
 
     transaction_processor: TransactionBatchProcessor<BankForks>,
+
+    /// Collected fee details
+    collector_fee_details: RwLock<CollectorFeeDetails>,
```

Comment:
`CollectorFeeDetails` probably doesn't need to be part of PartialEq.

`bank` still needs to read it right before distribution. But that happens when bank frozen. So arguably, nothing writes to it anymore, so mutex should be sufficient. I still wonder if there is use case when some threads want to read (say metrics or log) while bank is still progressing, which will be a case for RwLock.

 On the other hand, how much we gain by using mutex in this case?

Reply:
Probably not too much, but it is simple to swap out and access is mutally exclusive so it's the appropriate lock type it seems.

It's just another lock worker threads are contentious over per transaction batch.
We should measure impact of this additional using banking_bench, as was done in
 
We should be using the most efficient lock type for the access, which in this case is mutally exclusive. Honestly, it's worth considering atomics instead - we can use the `Relaxed` since we only care about the total accumulation and not value at a particular moment.

Reply:
Would think in multiple threading, RwLock would be less contentious than mutex, should the object to be read/write. This is not the case atm for CollectorFeeDetails, but hard to say it will never be, given all potential changes around fee/reward.

It is better to do bench with RwLock/Mutex/Atomic, I'll report back results here.

Reply:
`$ ./cargo nightly run --release --manifest-path banking-bench/Cargo.toml -- --write-lock-contention full  --skip-sanity`
- RwLock
```
[total_sent: 1152000, base_tx_count: 18432, txs_processed: 1107947, txs_landed: 1089515, total_us: 68826347, tx_total_us: 67682138]
{'name': 'banking_bench_total', 'median': '16737.78'}
{'name': 'banking_bench_tx_total', 'median': '17020.74'}
{'name': 'banking_bench_success_tx_total', 'median': '15829.91'}
```
- mutex
```
[total_sent: 1152000, base_tx_count: 18432, txs_processed: 1113143, txs_landed: 1094711, total_us: 70108159, tx_total_us: 68942982]
{'name': 'banking_bench_total', 'median': '16431.75'}
{'name': 'banking_bench_tx_total', 'median': '16709.46'}
{'name': 'banking_bench_success_tx_total', 'median': '15614.60'}
```
- atomic
```
[total_sent: 1152000, base_tx_count: 18432, txs_processed: 1110122, txs_landed: 1091690, total_us: 69012472, tx_total_us: 67828224]
{'name': 'banking_bench_total', 'median': '16692.63'}
{'name': 'banking_bench_tx_total', 'median': '16984.08'}
{'name': 'banking_bench_success_tx_total', 'median': '15818.73'}
```

`$./cargo nightly run --release --manifest-path banking-bench/Cargo.toml -- --write-lock-contention none  --skip-sanity`
- RwLock
```
[total_sent: 1152000, base_tx_count: 18432, txs_processed: 1151872, txs_landed: 1133440, total_us: 18858272, tx_total_us: 18295779]
{'name': 'banking_bench_total', 'median': '61087.25'}
{'name': 'banking_bench_tx_total', 'median': '62965.34'}
{'name': 'banking_bench_success_tx_total', 'median': '60103.07'}
```
- Mutex
```
[total_sent: 1152000, base_tx_count: 18432, txs_processed: 1148928, txs_landed: 1130496, total_us: 18840413, tx_total_us: 18268472]
{'name': 'banking_bench_total', 'median': '61145.16'}
{'name': 'banking_bench_tx_total', 'median': '63059.46'}
{'name': 'banking_bench_success_tx_total', 'median': '60003.78'}
```
- Atomic
```
[total_sent: 1152000, base_tx_count: 18432, txs_processed: 1148288, txs_landed: 1129856, total_us: 19163449, tx_total_us: 18582160]
{'name': 'banking_bench_total', 'median': '60114.44'}
{'name': 'banking_bench_tx_total', 'median': '61994.95'}
{'name': 'banking_bench_success_tx_total', 'median': '58958.91'}
```

There isn't significant perf advantage between locking methods. Not too sure about `AtomicU64`s, neutral between RwLock and Mutex, after all, they are really easy to swap.

Reply:
bench code: https://github.com/anza-xyz/agave/pull/232

Reply:
Great, thanks for running it. Figured it wouldn't have much impact, but good to always double check - any of the solutions seem fine to me.

Reply:
Let's keep it RwLock for now, unless there are other concerns.